### PR TITLE
Avoid UB on converting NaN to int in annotations

### PR DESCRIPTION
**Pull request by averkhaturau**
# Issue

Speed annotation may fail with converting NaN to integer if the same geometry point in duplicated in a route. 

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/thor/triplegbuilder.cc
Comment (SiarheiFedartsou) [inline src/thor/triplegbuilder.cc:331]: probably [`std::isnan`](https://en.cppreference.com/w/cpp/numeric/math/isnan) is more readable

**Review by SiarheiFedartsou — DISMISSED** [review]

**Review by SiarheiFedartsou — DISMISSED** [review]

**Comment by kevinkreiser** [comment]
Why does the duplicate occur? I thought the cutting logic was robust to this which means it has to be duplicated shape in OSM itself? If so that is an easy gruka test to write to actually get test coverage here. Shouldn't we add that quickly before merging?

**Review by danpat — CHANGES_REQUESTED** [review]
Can someone add a test for this case?  @kevinkreiser yeah, this is occuring when the underlying OSM data contains a zero-length, duplicate-coordinate node.

File: test/shape_attributes.cc
Comment (PureW) [inline test/shape_attributes.cc:86]: This test passes on master as well...

**Comment by kevinkreiser** [comment]
i wonder if this line is a contributing factor:

`double speed = (edge->length() * (tgt_pct - src_pct)) / edge_seconds;`

maybe speed is 0 if at the beginning or end of a map match we use 0% of an edge or maybe edge_seconds is 0. @averkhaturau i assume you ran into this sometime. did you happen to debug it to see the root cause of the nan?

File: test/shape_attributes.cc
Comment (danpat) [reply test/shape_attributes.cc:86]: @PureW the test will pass, but is `ubsan` triggered?

Comment (averkhaturau) [reply test/shape_attributes.cc:86]: I'm not sure how to test UB without ubsan, I'll enable it on CI

**Comment by averkhaturau** [comment]
> did you happen to debug it to see the root cause of the nan?

In short, if we have a duplicated coordinate in route geometry (we usually do in multileg routes), the speed is calculated as
`speed = length / time`
Both `length` and `time` are 0., so speed is set to `NaN`. Then it converts to integer in call of `set_speed(int)`, which is undefined behavior. Happily (or actually unhappily) clang converts NaN to integer 0 and nothing awful happens. ubsan catches it.

So, the reason is actually a duplicated coordinate.

**Comment by kevinkreiser** [comment]
I do not understand your comment about multileg routes! TripLegBuilder only works on one leg at a time, so surely multileg routes cannot cause this problem. I'm just wondering where the point duplication comes in. Is it in the base OSM data like that or do we have a bug elsewhere that is erroneously repeating a coordinate (ie when cutting the shape for the first or last edge of the route, also done in triplegbuilder). its good to handle it because OSM could have duplicate adjacent points (and im certain we dont weed that out at tile building time), but my main question is what is causing duplicated points and do we need to fix something more in a separate PR

File: test/shape_attributes.cc
Comment (kevinkreiser) [reply test/shape_attributes.cc:86]: i would be fine to just undo the fix in triplegbuilder, turn on ubsan locally, show that the sanitizer finds the issue via the unit test and post it to this PR. we shouldnt change the behavior of CI in this PR, iirc when we had that enabled we either got randomly OOM killed or the build took forever

**Comment by merkispavel** [comment]
> I do not understand your comment about multileg routes! TripLegBuilder only works on one leg at a time, so surely multileg routes cannot cause this problem. I'm just wondering where the point duplication comes in. Is it in the base OSM data like that or do we have a bug elsewhere that is erroneously repeating a coordinate (ie when cutting the shape for the first or last edge of the route, also done in triplegbuilder)? its good to handle it because OSM could have duplicate adjacent points (and im certain we dont weed that out at tile building time), but my main question is what is causing duplicated points and do we need to fix something more in a separate PR

@kevinkreiser as far as i see we do have request that reproduces the issue. I'll come back to you soon and let you know if we need one more fix somewhere else. But I believe this activity is not a blocker for this PR

File: test/shape_attributes.cc
Comment (averkhaturau) [reply test/shape_attributes.cc:86]: In my opinion, CI should test valhalla for UB, it would help to avoid UB issues in future, It's not the only issue found with ubsan.

**Comment by averkhaturau** [comment]
This PR is a hot workaround of UB. Duplicated point is a separate issue.

**Comment by kevinkreiser** [comment]
yeah im not trying to block the PR. let me reiterate my points more concisely:

> its good to handle it because OSM could have duplicate adjacent points
^^ we should ship this to protect us from NaN regardless of the cause

> what is causing duplicated points and do we need to fix something more in a separate PR
^^ lets investigate the cause of the duplicate adjacent shape points in a separate issue/PR

> we shouldnt change the behavior of CI in this PR
^^ revert the changes to CI

**Comment by averkhaturau** [comment]
@kevinkreiser , I reverted CI changes. Should I revert a test as well?

**Review by merkispavel — APPROVED** [review]
Looks good

**Review by danpat — APPROVED** [review]

**Comment by danpat** [comment]
Duplicated coordinates can _easily_ exist in the original OSM data - I don't think there's any need to look for another cause here.  A simple scan of OSM I'm sure will expose thousands, if not millions, of cases of this.  Valhalla needs to be robust against that, and as @averkhaturau pointed out, we've been lucky that it just so happens that NaN->0 for our serialization in most cases.

**Pull request merged**
